netket/graph/lattice.py

```diff
@@ -491,37 +445,48 @@ def basis_coords(self) -> CoordT:
 
     # Site lookup
     # ------------------------------------------------------------------------
+
+    class InvalidSiteError(Exception):
+        pass
+
+    class InvalidWaveVectorError(Exception):
+        pass
+
     def _to_integer_position(self, positions: PositionT) -> Array:
-        frac_positions = _np.matmul(positions, self._inv_dims) % 1
-        return _np.around(frac_positions * 10 ** tol_digits).astype(int) % (
-            10 ** tol_digits
-        )
+        frac_positions = _np.matmul(positions, self._inv_dims)
+        return comparable_periodic(frac_positions, self.pbc)
 
     @staticmethod
     def _get_id_from_dict(
         dict: Dict[HashableArray, int], key: Array
     ) -> Union[int, Array]:
-        if key.ndim == 1:
-            return dict.get(HashableArray(key), None)
-        elif key.ndim == 2:
-            return _np.array([dict.get(HashableArray(k), None) for k in key])
-        else:
-            raise ValueError("Input needs to be rank 1 or rank 2 array")
+        try:
+            if key.ndim == 1:
+                return dict[HashableArray(key)]
+            elif key.ndim == 2:
+                return _np.array([dict[HashableArray(k)] for k in key])
+            else:
+                raise ValueError("Input needs to be rank 1 or rank 2 array")
+        except KeyError:
+            raise Lattice.InvalidSiteError(
+                "Some coordinates do not correspond to a valid lattice site"
+            )
 
     def id_from_position(self, position: PositionT) -> Union[int, Array]:
```

Comment:
I'm wondering if we shouldn't rename this to `index_from_position` rather than `id`, as it would be clearer. @femtobit

Reply:
I did choose `id` over `index` because `index` can be used for many different things here (e.g., index of a site in `site_offsets` [we call this `label`], index of a spatial direction in position or basis coordinate). So referring to `site index` could be more ambiguous than `site id` (which we should make sure to only ever use to mean the one unique integer identifying the site in the context of lattice).

I am not 100%-sure about the naming, though. Do you think index would equally unambiguous in practice?

Reply:
index and id are pretty interchangeable. Maybe index reminds more of an array and id more of a dict.
We are using a dict after all, so maybe that's fine.
But I find it confusing..

Reply:
What matter most is that we pick one and stick with it. (Both are more specific than `site` as before.) Maybe make a separate issue as I wouldn't change that in this PR?

Reply:
ok